### PR TITLE
Adding CouchDB properties for config

### LINKED ISSUE #943: couch parameters in pac4j-config
I'm planning to make a pull request to add a `CouchAuthenticatorBuilder` in pac4j-config, please comment

### PR DESCRIPTION
**Pull request by eroux**
This fixes #943. I think it's ready for review, I can add the inmemory authenticator after comments are received (and #944 is merged)

### CODE REVIEW

**Comment by eroux** [comment]
This one is ready too, comments/review welcome!

File: pac4j-config/src/main/java/org/pac4j/config/builder/CouchAuthenticatorBuilder.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+public class CouchAuthenticatorBuilder  extends AbstractBuilder {
+
+    public CouchAuthenticatorBuilder(final Map<String, String> properties) {
+        super(properties);
+    }
+
+    public void tryBuildDbAuthenticator(final Map<String, Authenticator> authenticators, final Map<String, PasswordEncoder> encoders) {
+        for (int i = 0; i <= MAX_NUM_AUTHENTICATORS; i++) {
+            if (containsProperty(COUCH_URL, i) || containsProperty(COUCH_DATABASENAME, i)) {
```
Comment (leleuj) [inline pac4j-config/src/main/java/org/pac4j/config/builder/CouchAuthenticatorBuilder.java]: Reading the source code above, I have the feeling both properties are required.

File: pac4j-config/src/main/java/org/pac4j/config/builder/CouchAuthenticatorBuilder.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                final Builder httpClientBuilder = new StdHttpClient.Builder();
+                if (containsProperty(COUCH_USERNAME, i)) {
+                    httpClientBuilder.username(getProperty(COUCH_USERNAME, i));
+                }
+                if (containsProperty(COUCH_PASSWORD, i)) {
+                    httpClientBuilder.password(getProperty(COUCH_PASSWORD, i));
+                }
+                if (containsProperty(COUCH_URL, i)) {
```
Comment (leleuj) [inline pac4j-config/src/main/java/org/pac4j/config/builder/CouchAuthenticatorBuilder.java]: I think this property should be mandatory, so no `if`/`else` clause.

File: pac4j-config/src/main/java/org/pac4j/config/builder/CouchAuthenticatorBuilder.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                    httpClientBuilder.password(getProperty(COUCH_URL, i));
+                } else {
+                    throw new TechnicalException("missing "+COUCH_URL+" property in CouchDB configuration");
+                }
+                final HttpClient httpClient = httpClientBuilder.build();
+                final CouchDbInstance dbInstance = new StdCouchDbInstance(httpClient);
+                if (!containsProperty(COUCH_DATABASENAME, i)) {
```
Comment (leleuj) [inline pac4j-config/src/main/java/org/pac4j/config/builder/CouchAuthenticatorBuilder.java]: I think this property should be mandatory, so no `if`/`else` clause.

File: pac4j-config/src/main/java/org/pac4j/config/builder/CouchAuthenticatorBuilder.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            if (containsProperty(COUCH_URL, i) || containsProperty(COUCH_DATABASENAME, i)) {
+                final Builder httpClientBuilder = new StdHttpClient.Builder();
+                if (containsProperty(COUCH_USERNAME, i)) {
+                    httpClientBuilder.username(getProperty(COUCH_USERNAME, i));
+                }
+                if (containsProperty(COUCH_PASSWORD, i)) {
+                    httpClientBuilder.password(getProperty(COUCH_PASSWORD, i));
+                }
+                if (containsProperty(COUCH_URL, i)) {
+                    httpClientBuilder.password(getProperty(COUCH_URL, i));
```
Comment (leleuj) [inline pac4j-config/src/main/java/org/pac4j/config/builder/CouchAuthenticatorBuilder.java]: `password` method on the URL?

**Comment by leleuj** [comment]
A few comments. Overall, it looks good...

**Comment by eroux** [comment]
Thanks for the review! comments addressed

**Comment by leleuj** [comment]
In fact, there is a core issue, your `CouchAuthenticatorBuilder` is not used in the `PropertiesConfigFactory`: it will never work. It should be something like the LDAP or SQL support. And it would be better to upgrade the test case to confirm the behaviour.

**Comment by leleuj** [comment]
@eroux any update on this?

**Comment by eroux** [comment]
Not at the moment sorry, I'll try to finish that soon

**Comment by leleuj** [comment]
No problem. I thought you may have missed my feedback. Thus the reminder...

**Comment by leleuj** [comment]
I'm not sure to understand why the `CouchProfileService` class has changed, the whole indentation change makes impossible to know what has really changed.
But I still don't see the required change on the `PropertiesConfigFactory` component.

**Comment by eroux** [comment]
oh sorry, that's clearly a mistake, forget it!

**Comment by leleuj** [comment]
The idea behind the original issue was good (https://github.com/pac4j/pac4j/issues/943), even if the PR has not fully reached the final goal. Are you sure you want to close it and give up on the properties configuration of the CouchDB authentication?

**Comment by eroux** [comment]
Well, maybe I'll reopen it later, but I'm not at a stage where I can fully test it and there doesn't seem to be any possible unit testing possible (by design), so I'm giving up for the moment, maybe I'll reopen it later if I'm able to test it properly. But I think #958 and #959 are much more important

**Comment by leleuj** [comment]
OK. No problem. Let's close it for now.

**Pull request closed without merging**